trie/iterator.go

```diff
@@ -763,3 +767,291 @@ func (it *unionIterator) Error() error {
+	start []byte,
+) *merkleTreeIterator {
+	it := &merkleTreeIterator{nodeBlobFromTree: nodeBlobFromTree, nodeBlobToIteratorNode: nodeBlobToIteratorNode}
+	var rootNode merkleTreeIteratorNode
+	if rootNode, it.err = nodeBlobToIteratorNode(root, nodeBlobFromTree(root)); it.err == nil {
+		it.stack = []merkleTreeIteratorNode{rootNode}
+		it.seek(start)
+	}
+	return it
+}
+
+func (it *merkleTreeIterator) seek(path []byte) {
+	if len(path) == 0 {
+		return
+	}
+
+	for _, p := range path {
+		if it.err != nil {
+			return
+		}
+		if parent, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorParentNode); ok {
+			if child := it.resolveNode(parent.children[p]); child != nil {
+				it.stack = append(it.stack, child)
+				it.path = append(it.path, p)
+				continue
+			}
+		}
+	}
+
+	// When using the Next function, it always moves to the next node.
+	// Therefore, it sets the previous node of the node retrieved by seek to the last visited node.
+	if it.path[len(it.path)-1] == right {
+		it.path[len(it.path)-1] = left
+		it.stack[len(it.stack)-1] = it.resolveNode(it.parentOfLastNode().children[left])
+	} else {
+		it.path = it.path[:len(it.path)-1]
+		it.stack = it.stack[:len(it.stack)-1]
+	}
+}
+
+func (it *merkleTreeIterator) Next(bool) bool {
+	if it.err != nil {
+		return false
+	}
+	if it.stack == nil { // end of iterator traversal
+		return false
+	}
+	if it.path == nil { // first visit. Starting from the root
+		it.path = []byte{}
+		return true
+	}
+	switch last := it.stack[len(it.stack)-1].(type) {
+	case *merkleTreeIteratorParentNode:
+		// find the next node by preorder traversal. The children must have at least one not null.
+		for p, hash := range last.children {
+			if child := it.resolveNode(hash); child != nil {
+				it.path = append(it.path, byte(p))
+				it.stack = append(it.stack, child)
+				return true
+			}
+		}
+		it.err = errors.New("not all child nodes exist")
+		return false
+	case *merkleTreeIteratorLeafNode:
+		for { // go around an infinite loop until the next visitable node is found.
+			switch lastPathIndex := len(it.path) - 1; it.path[lastPathIndex] {
+			case 0: // left visited. go right
+				if rightNode := it.resolveNode(it.parentOfLastNode().children[right]); rightNode != nil {
+					it.path[lastPathIndex] = right
+					it.stack[len(it.stack)-1] = rightNode
+					return true
+				}
+				fallthrough // right does not exist. go up (processing in the bottom case code)
+			case 1: // right visited. go up
+				it.path = it.path[:len(it.path)-1]
+				it.stack = it.stack[:len(it.stack)-1]
+				if len(it.path) == 0 { // no visitable node exists
+					return false
+				}
+			}
+		}
+	}
+	return false
+}
+
+func (it *merkleTreeIterator) Error() error      { return it.err }
+func (it *merkleTreeIterator) Hash() common.Hash { return it.stack[len(it.stack)-1].Hash() }
+
+func (it *merkleTreeIterator) Parent() common.Hash {
+	switch l := len(it.stack); l {
+	case 0:
+		return common.Hash{}
+	case 1:
+		return it.stack[0].Hash()
```

Comment:
```suggestion
	case left:
		return common.Hash{}
	case right:
		return it.stack[0].Hash()
```

Reply:
That's not a tree path, that's a tree depth.